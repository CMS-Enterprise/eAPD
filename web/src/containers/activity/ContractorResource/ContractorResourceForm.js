import { FormLabel, TextField } from '@cmsgov/design-system-core';
import PropTypes from 'prop-types';
import React, { Fragment, useMemo, useState } from 'react';
import { connect } from 'react-redux';

import Choice from '../../../components/Choice';
import DateField from '../../../components/DateField';
import DollarField from '../../../components/DollarField';
import Dollars from '../../../components/Dollars';
import TextArea from '../../../components/TextArea';
import NumberField from '../../../components/NumberField';

import {
  setContractorDescription,
  setContractorEndDate,
  setContractorIsHourly,
  setContractorName,
  setContractorStartDate,
  setContractorTotalCost,
  setContractorCostForYear,
  setContractorNumberOfHoursForYear,
  setContractorHourlyRateForYear
} from '../../../actions/editActivity';

const ContractorResourceForm = ({
  activityIndex,
  index,
  item: { description, end, hourly, key, name, start, totalCost, years },
  setDescription,
  setEndDate,
  setIsHourly,
  setName,
  setStartDate,
  setTotalCost,
  setCostForYear,
  setNumberOfHoursForYear,
  setHourlyRateForYear
}) => {
  const apdFFYs = useMemo(() => Object.keys(years), [years]);
  const [nonHourlyCost, setNonHourlyCost] = useState(
    apdFFYs.reduce(
      (o, ffy) => ({ ...o, [ffy]: !hourly.useHourly ? years[ffy] : '' }),
      {}
    )
  );

  const getHandler = action => ({ target: { value } }) => {
    action(activityIndex, index, value);
  };

  const getDateHandler = action => (_, dateStr) => {
    action(activityIndex, index, dateStr);
  };

  const setHourlyOff = () => {
    setIsHourly(activityIndex, index, false);

    // set cost to non-hourly cost saved from last time the radio button was switched (if any)
    apdFFYs.forEach(ffy => {
      if (nonHourlyCost[ffy]) {
        setCostForYear(activityIndex, index, ffy, nonHourlyCost[ffy]);
      }
    });
  };

  const setHourlyOn = () => {
    // save non-hourly cost in case the radio button is switched back to non-hourly
    setNonHourlyCost(
      apdFFYs.reduce((o, ffy) => ({ ...o, [ffy]: years[ffy] }), {})
    );

    setIsHourly(activityIndex, index, true);

    // set cost to hours x rate
    apdFFYs.forEach(ffy => {
      if (hourly.data[ffy].hours && hourly.data[ffy].rate) {
        setCostForYear(
          activityIndex,
          index,
          ffy,
          hourly.data[ffy].hours * hourly.data[ffy].rate
        );
      }
    });
  };

  const getHandlerForYearlyCost = year => ({ target: { value } }) => {
    setCostForYear(activityIndex, index, year, value);
  };

  const getHandlerForYearlyHours = year => ({ target: { value } }) => {
    setNumberOfHoursForYear(activityIndex, index, year, value);
  };

  const getHandlerForYearlyHourlyRate = year => ({ target: { value } }) => {
    setHourlyRateForYear(activityIndex, index, year, value);
  };

  return (
    <Fragment>
      <TextField
        autoFocus
        label="Contractor Name"
        name="contractor-name"
        value={name}
        onChange={getHandler(setName)}
      />
      <TextArea
        label="Description of Services"
        name="contractor-description"
        rows={5}
        value={description}
        onChange={getHandler(setDescription)}
      />
      <FormLabel>Contract Term</FormLabel>
      <div className="ds-c-choice__checkedChild ds-u-padding-y--0">
        <DateField
          label="Start date"
          value={start}
          onChange={getDateHandler(setStartDate)}
        />
        <DateField
          label="End date"
          hint=""
          value={end}
          onChange={getDateHandler(setEndDate)}
        />
      </div>
      <DollarField
        label="Total Contract Cost"
        name="contractor-total-cost"
        size="medium"
        value={totalCost}
        onChange={getHandler(setTotalCost)}
      />

      <fieldset className="ds-c-fieldset">
        <legend className="ds-c-label">This is an hourly resource</legend>
        <Choice
          type="radio"
          name={`apd-activity-contractor-hourly-${key}-no`}
          value="no"
          checked={!hourly.useHourly}
          onChange={setHourlyOff}
        >
          No
        </Choice>
        <Choice
          type="radio"
          name={`apd-activity-contractor-hourly-${key}-yes`}
          value="yes"
          checked={hourly.useHourly}
          onChange={setHourlyOn}
          checkedChildren={
            <div className="ds-c-choice__checkedChild">
              {apdFFYs.map(ffy => (
                <Fragment key={ffy}>
                  <FormLabel>FFY {ffy}</FormLabel>
                  <div className="ds-l-row ds-u-padding-left--2">
                    <NumberField
                      label="Number of hours"
                      name={`contractor-num-hours-ffy-${ffy}`}
                      labelClassName="ds-u-margin-top--1"
                      size="medium"
                      value={hourly.data[ffy].hours}
                      onChange={getHandlerForYearlyHours(ffy)}
                    />
                    <DollarField
                      className="ds-u-margin-left--1"
                      label="Hourly rate"
                      name={`contractor-hourly-rate-ffy-${ffy}`}
                      labelClassName="ds-u-margin-top--1"
                      size="medium"
                      value={hourly.data[ffy].rate}
                      onChange={getHandlerForYearlyHourlyRate(ffy)}
                    />
                  </div>
                </Fragment>
              ))}
            </div>
          }
        >
          Yes
        </Choice>
      </fieldset>

      {hourly.useHourly ? (
        <p className="ds-u-margin-bottom--0">
          {apdFFYs.map(ffy => (
            <Fragment key={ffy}>
              <FormLabel>FFY {ffy} Cost</FormLabel>
              <Dollars>{years[ffy]}</Dollars>
            </Fragment>
          ))}
        </p>
      ) : (
        apdFFYs.map(ffy => (
          <DollarField
            key={ffy}
            label={`FFY ${ffy} Cost`}
            name={`contractor-cost-ffy-${ffy}`}
            disabled={hourly.useHourly}
            size="medium"
            value={years[ffy]}
            onChange={getHandlerForYearlyCost(ffy)}
          />
        ))
      )}
    </Fragment>
  );
};

ContractorResourceForm.propTypes = {
  activityIndex: PropTypes.number.isRequired,
  index: PropTypes.number.isRequired,
  item: PropTypes.shape({
    description: PropTypes.string,
    end: PropTypes.string,
    hourly: PropTypes.object,
    key: PropTypes.string,
    name: PropTypes.string,
    start: PropTypes.string,
    totalCost: PropTypes.number,
    years: PropTypes.object
  }).isRequired,
  setDescription: PropTypes.func.isRequired,
  setEndDate: PropTypes.func.isRequired,
  setIsHourly: PropTypes.func.isRequired,
  setName: PropTypes.func.isRequired,
  setStartDate: PropTypes.func.isRequired,
  setTotalCost: PropTypes.func.isRequired,
  setCostForYear: PropTypes.func.isRequired,
  setNumberOfHoursForYear: PropTypes.func.isRequired,
  setHourlyRateForYear: PropTypes.func.isRequired
};

const mapDispatchToProps = {
  setDescription: setContractorDescription,
  setEndDate: setContractorEndDate,
  setIsHourly: setContractorIsHourly,
  setName: setContractorName,
  setStartDate: setContractorStartDate,
  setTotalCost: setContractorTotalCost,
  setCostForYear: setContractorCostForYear,
  setNumberOfHoursForYear: setContractorNumberOfHoursForYear,
  setHourlyRateForYear: setContractorHourlyRateForYear
};

export default connect(null, mapDispatchToProps)(ContractorResourceForm);

export { ContractorResourceForm as plain, mapDispatchToProps };
