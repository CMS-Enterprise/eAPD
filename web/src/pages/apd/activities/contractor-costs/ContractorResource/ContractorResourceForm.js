import { FormLabel, TextField, ChoiceList } from '@cmsgov/design-system';
import PropTypes from 'prop-types';
import React, {
  Fragment,
  forwardRef,
  useMemo,
  useReducer,
  useEffect
} from 'react';
import { useForm, Controller } from 'react-hook-form';
import { joiResolver } from '@hookform/resolvers/joi';
import { connect } from 'react-redux';

import DateField from '../../../../../components/DateField';
import DollarField from '../../../../../components/DollarField';
import Dollars from '../../../../../components/Dollars';
import NumberField from '../../../../../components/NumberField';
import RichText from '../../../../../components/RichText';

import validationSchema from '../../../../../static/schemas/privateContractor';
import { saveContractor as actualSaveContractor } from '../../../../../actions/editActivity';

const getCheckedValue = value => {
  if (value !== null) {
    if (value === true) return 'yes';
    if (value === false) return 'no';
    return value;
  }
  return null;
};

const ContractorResourceForm = forwardRef(
  ({ activityIndex, index, item, saveContractor, setFormValid }, ref) => {
    ContractorResourceForm.displayName = 'ContractorResourceForm';
    const {
      handleSubmit,
      control,
      getFieldState,
      trigger,
      formState: { errors, isValid },
      resetField: resetFieldErrors
    } = useForm({
      defaultValues: {
        name: item.name,
        description: item.description,
        start: item.start,
        end: item.end,
        totalCost: item.totalCost,
        useHourly: getCheckedValue(item.useHourly),
        years: item.years,
        hourly: item.hourly
      },
      mode: 'onBlur',
      reValidateMode: 'onBlur',
      resolver: joiResolver(validationSchema)
    });

    useEffect(() => {
      setFormValid(isValid);
    }, [isValid]); // eslint-disable-line react-hooks/exhaustive-deps

    const initialState = item;

    function reducer(state, action) {
      switch (action.type) {
        case 'updateField':
          return {
            ...state,
            [action.field]: action.value
          };
        case 'setHourly':
          return {
            ...state,
            useHourly: action.value
          };
        case 'updateHourlyHours':
          return {
            ...state,
            hourly: {
              ...state.hourly,
              [action.year]: {
                ...state.hourly[action.year],
                hours: action.value
              }
            }
          };
        case 'updateHourlyRate':
          return {
            ...state,
            hourly: {
              ...state.hourly,
              [action.year]: {
                ...state.hourly[action.year],
                rate: action.value
              }
            }
          };
        case 'updateYearCost':
          return {
            ...state,
            years: {
              ...state.years,
              [action.year]: action.value
            }
          };
        default:
          throw new Error(
            'Unrecognized action type provided to ContractorResourceForm reducer'
          );
      }
    }

    const [state, dispatch] = useReducer(reducer, initialState);

    const apdFFYs = useMemo(() => Object.keys(state.years), [state.years]);

    const handleNameChange = e => {
      dispatch({
        type: 'updateField',
        field: 'name',
        value: e.target.value
      });
    };

    const handleDescriptionChange = value => {
      dispatch({
        type: 'updateField',
        field: 'description',
        value
      });
    };

    const handleTotalCostChange = e => {
      dispatch({
        type: 'updateField',
        field: 'totalCost',
        value: e.target.value
      });
    };

    const handleDateChange = (field, value) => {
      dispatch({
        type: 'updateField',
        field,
        value
      });
    };

    const handleUseHourlyChange = e => {
      dispatch({ type: 'setHourly', value: e.target.value });
    };

    const handleHourlyHoursChange = (ffy, e) => {
      dispatch({
        type: 'updateHourlyHours',
        year: ffy,
        value: e.target.value
      });
      dispatch({
        type: 'updateYearCost',
        year: ffy,
        value: e.target.value * state.hourly[ffy].rate
      });
    };

    const handleHourlyRateChange = (ffy, e) => {
      dispatch({
        type: 'updateHourlyRate',
        year: ffy,
        value: e.target.value
      });
      dispatch({
        type: 'updateYearCost',
        year: ffy,
        value: state.hourly[ffy].hours * e.target.value
      });
    };

    const handleYearCostChange = (ffy, e) => {
      dispatch({
        type: 'updateYearCost',
        year: ffy,
        value: e.target.value
      });
    };

    const onSubmit = e => {
      e.preventDefault();
      saveContractor(activityIndex, index, state);
      handleSubmit(e);
    };

    return (
      <form index={index} onSubmit={onSubmit} aria-label="form">
        <Controller
          name="name"
          control={control}
          render={({ field: { onChange, ...props } }) => (
            <TextField
              {...props}
              label="Private Contractor or Vendor Name"
              hint="Provide the name of the private contractor or vendor. For planned procurements, generalize by resource name. For example, Computer Resources/TBD."
              labelClassName="full-width-label"
              className="remove-clearfix"
              onChange={e => {
                handleNameChange(e);
                onChange(e);
              }}
              errorMessage={errors?.name?.message}
              errorPlacement="bottom"
            />
          )}
        />
        <FormLabel
          className="full-width-label"
          fieldId={`contractor-description-field-${index}`}
        >
          Procurement Methodology and Description of Services
          <span className="ds-c-field__hint ds-u-margin--0">
            Explain the procurement process for the contractor and the scope of
            their work. Provide justification for any non-competitive
            procurements.
          </span>
        </FormLabel>
        <Controller
          name="description"
          control={control}
          render={({ field: { onChange, onBlur } }) => (
            <RichText
              name="contractor-description"
              id={`contractor-description-field-${index}`}
              content={state.description}
              onSync={html => {
                handleDescriptionChange(html);
                onChange(html);
              }}
              onBlur={onBlur}
              editorClassName="rte-textarea-1"
            />
          )}
        />
        {errors?.description && (
          <span
            className="ds-c-inline-error ds-c-field__error-message"
            role="alert"
          >
            {errors.description.message}
          </span>
        )}
        <FormLabel>Full Contract Term</FormLabel>
        <span className="ds-c-field__hint">
          Provide the total length of the contract, including any option years.
          Contract term may extend beyond the FFY(s) included in this APD.
        </span>
        <div className="ds-c-choice__checkedChild ds-u-padding-y--0">
          <Controller
            name="start"
            control={control}
            render={({ field: { onBlur, onChange, name, value } }) => (
              <DateField
                name={name}
                label="Contract start date"
                value={value}
                onChange={(e, dateStr) => {
                  handleDateChange('start', dateStr);
                  onChange(dateStr);
                }}
                onComponentBlur={(e, dateStr) => {
                  if (getFieldState('end').isTouched) trigger('end');
                  onBlur(dateStr);
                }}
                errorMessage={errors?.start?.message}
              />
            )}
          />
          <Controller
            name="end"
            control={control}
            render={({ field: { onBlur, onChange, name, value } }) => {
              return (
                <DateField
                  name={name}
                  label="Contract end date"
                  value={value}
                  onChange={(e, dateStr) => {
                    handleDateChange('end', dateStr);
                    onChange(dateStr);
                  }}
                  onComponentBlur={(e, dateStr) => {
                    if (getFieldState('start').isTouched) trigger('start');
                    onBlur(dateStr);
                  }}
                  errorMessage={errors?.end?.message}
                />
              );
            }}
          />
        </div>
        <Controller
          name="totalCost"
          control={control}
          render={({ field: { onChange, onBlur, name, value } }) => (
            <DollarField
              name={name}
              label="Total Contract Cost"
              size="medium"
              hint="Provide the total not to exceed amounts of the contract, including costs for the option years. This is not the amount you are requesting for the FFYs and will not be added to your FFY requests."
              labelClassName="full-width-label"
              value={value}
              onBlur={onBlur}
              onChange={e => {
                handleTotalCostChange(e);
                onChange(e);
              }}
              errorMessage={errors?.totalCost?.message}
              errorPlacement="bottom"
            />
          )}
        />
        <Controller
          name="useHourly"
          control={control}
          render={({
            field: { name, onChange, onBlur: useHourlyOnBlur, value }
          }) => (
            <ChoiceList
              label="This is an hourly resource"
              name={name}
              choices={[
                {
                  label: 'Yes',
                  value: 'yes',
                  checked: value === 'yes',
                  checkedChildren: (
                    <div className="ds-c-choice__checkedChild">
                      {apdFFYs.map(ffy => (
                        <Fragment key={ffy}>
                          <FormLabel>FFY {ffy}</FormLabel>
                          <div className="ds-l-row ds-u-padding-left--2">
                            <Controller
                              key={`${ffy}-hours`}
                              name={`hourly.${ffy}.hours`}
                              control={control}
                              render={({
                                field: { onChange: hoursOnChange, ...props }
                              }) => (
                                <NumberField
                                  {...props}
                                  label="Number of hours"
                                  labelClassName="ds-u-margin-top--1"
                                  size="medium"
                                  onChange={e => {
                                    handleHourlyHoursChange(ffy, e);
                                    hoursOnChange(e);
                                  }}
                                />
                              )}
                            />
                            <Controller
                              key={`${ffy}-rate`}
                              name={`hourly.${ffy}.rate`}
                              control={control}
                              render={({
                                field: {
                                  onChange: rateOnChange,
                                  onBlur: rateOnBlur,
                                  name,
                                  value
                                }
                              }) => (
                                <DollarField
                                  name={name}
                                  value={value}
                                  className="ds-u-margin-left--1"
                                  label="Hourly rate"
                                  labelClassName="ds-u-margin-top--1"
                                  size="medium"
                                  onBlur={rateOnBlur}
                                  onChange={e => {
                                    handleHourlyRateChange(ffy, e);
                                    rateOnChange(e);
                                  }}
                                />
                              )}
                            />
                          </div>
                          <div>
                            <Fragment>
                              {errors?.hourly && errors.hourly[ffy]?.hours && (
                                <span
                                  className="ds-c-inline-error ds-c-field__error-message"
                                  role="alert"
                                >
                                  {errors.hourly[ffy]?.hours?.message}
                                </span>
                              )}
                              {errors?.hourly && errors.hourly[ffy]?.rate && (
                                <span
                                  className="ds-c-inline-error ds-c-field__error-message"
                                  role="alert"
                                >
                                  {errors.hourly[ffy]?.rate?.message}
                                </span>
                              )}
                            </Fragment>
                          </div>
                        </Fragment>
                      ))}
                    </div>
                  )
                },
                {
                  label: 'No',
                  value: 'no',
                  checked: value === 'no'
                }
              ]}
              type="radio"
              onChange={e => {
                resetFieldErrors('hourly');
                resetFieldErrors('years');
                handleUseHourlyChange(e);
                onChange(e);
              }}
              onBlur={useHourlyOnBlur}
              onComponentBlur={useHourlyOnBlur}
              errorMessage={errors?.useHourly?.message}
              errorPlacement="bottom"
            />
          )}
        />
        {state.useHourly === null ||
        state.useHourly === true ||
        state.useHourly === 'yes' ? (
          <p className="ds-u-margin-bottom--0">
            {apdFFYs.map(ffy => (
              <div key={ffy}>
                <FormLabel>FFY {ffy} Cost</FormLabel>
                <Dollars>{state.years[ffy]}</Dollars>
              </div>
            ))}
          </p>
        ) : (
          <p className="ds-u-margin-bottom--0">
            {apdFFYs.map(ffy => (
              <Controller
                key={ffy}
                name={`years.${ffy}`}
                control={control}
                render={({ field: { onChange, onBlur, name, value } }) => (
                  <DollarField
                    name={name}
                    value={value}
                    label={`FFY ${ffy} Cost`}
                    size="medium"
                    onBlur={onBlur}
                    onChange={e => {
                      handleYearCostChange(ffy, e);
                      onChange(e);
                    }}
                    errorMessage={errors?.years?.[ffy]?.message}
                    errorPlacement="bottom"
                  />
                )}
              />
            ))}
          </p>
        )}
        <input
          className="ds-u-visibility--hidden"
          type="submit"
          ref={ref}
          hidden
        />
      </form>
    );
  }
);

ContractorResourceForm.propTypes = {
  activityIndex: PropTypes.number.isRequired,
  index: PropTypes.number.isRequired,
  item: PropTypes.shape({
    key: PropTypes.string,
    name: PropTypes.string,
    description: PropTypes.string,
    start: PropTypes.string,
    end: PropTypes.string,
    totalCost: PropTypes.number,
    useHourly: PropTypes.bool,
    hourly: PropTypes.object,
    years: PropTypes.object
  }).isRequired,
  saveContractor: PropTypes.func.isRequired,
  setFormValid: PropTypes.func.isRequired
};

const mapDispatchToProps = {
  saveContractor: actualSaveContractor
};

export default connect(null, mapDispatchToProps, null, { forwardRef: true })(
  ContractorResourceForm
);

export { ContractorResourceForm as plain, mapDispatchToProps };
