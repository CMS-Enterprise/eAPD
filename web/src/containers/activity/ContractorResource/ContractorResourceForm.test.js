import { shallow } from 'enzyme';
import React from 'react';

import { plain as ContractorForm } from './ContractorResourceForm';

describe('the ContractorResourceForm component', () => {
  const props = {
    activityIndex: 43,
    index: 1,
    item: {
      description: 'They cleaned up the latrines after the Battle of Hastings',
      end: '1066-10-15',
      hourly: {
        useHourly: false,
        data: {
          '1066': {
            hours: 10,
            rate: 100
          },
          '1067': {
            hours: 20,
            rate: 200
          }
        }
      },
      id: 'contractor 1',
      key: 'key 1',
      name: 'Honey Dipper Dan',
      start: '1066-10-14',
      totalCost: 12345,
      years: {
        '1066': 300,
        '1067': 400
      }
    },
    setDescription: jest.fn(),
    setEndDate: jest.fn(),
    setIsHourly: jest.fn(),
    setName: jest.fn(),
    setStartDate: jest.fn(),
    setTotalCost: jest.fn(),
    setCostForYear: jest.fn(),
    setNumberOfHoursForYear: jest.fn(),
    setHourlyRateForYear: jest.fn()
  };

  beforeEach(() => {
    props.setDescription.mockClear();
    props.setEndDate.mockClear();
    props.setIsHourly.mockClear();
    props.setName.mockClear();
    props.setStartDate.mockClear();
    props.setTotalCost.mockClear();
    props.setCostForYear.mockClear();
    props.setNumberOfHoursForYear.mockClear();
    props.setHourlyRateForYear.mockClear();
  });

  test('renders correctly', () => {
    const component = shallow(<ContractorForm {...props} />);
    expect(component).toMatchSnapshot();
  });

  test('renders if use hourly data is selected', () => {
    props.item.hourly.useHourly = true;
    const component = shallow(<ContractorForm {...props} />);
    expect(component).toMatchSnapshot();
  });

  test('handles changing the contractor name', () => {
    const component = shallow(<ContractorForm {...props} />);
    component
      .findWhere(c => c.prop('name') === 'contractor-name')
      .simulate('change', { target: { value: 'new value' } });

    expect(props.setName).toHaveBeenCalledWith(43, 1, 'new value');
  });

  test('handles changing the contractor description', () => {
    const component = shallow(<ContractorForm {...props} />);
    component
      .findWhere(c => c.prop('name') === 'contractor-description')
      .simulate('change', { target: { value: 'new value' } });

    expect(props.setDescription).toHaveBeenCalledWith(43, 1, 'new value');
  });

  test('handles changing the contractor total cost', () => {
    const component = shallow(<ContractorForm {...props} />);
    component
      .findWhere(c => c.prop('name') === 'contractor-total-cost')
      .simulate('change', { target: { value: 'new value' } });

    expect(props.setTotalCost).toHaveBeenCalledWith(43, 1, 'new value');
  });

  test('handles changing the contractor start date', () => {
    const component = shallow(<ContractorForm {...props} />);
    component
      .findWhere(c => c.name() === 'DateField' && c.prop('label') === 'Start')
      .simulate('change', 'ignored stuff', 'new start date');

    expect(props.setStartDate).toHaveBeenCalledWith(43, 1, 'new start date');
  });

  test('handles changing the contractor end date', () => {
    const component = shallow(<ContractorForm {...props} />);
    component
      .findWhere(c => c.name() === 'DateField' && c.prop('label') === 'End')
      .simulate('change', 'ignored stuff', 'new end date');

    expect(props.setEndDate).toHaveBeenCalledWith(43, 1, 'new end date');
  });

  test('handles toggling hourly resource off', () => {
    const component = shallow(<ContractorForm {...props} />);
    component
      .findWhere(
        c => c.name() === 'ChoiceComponent' && c.prop('value') === 'no'
      )
      .simulate('change');

    expect(props.setIsHourly).toHaveBeenCalledWith(43, 1, false);
  });

  test('handles toggling hourly resource on', () => {
    const component = shallow(<ContractorForm {...props} />);
    component
      .findWhere(
        c => c.name() === 'ChoiceComponent' && c.prop('value') === 'yes'
      )
      .simulate('change');

    expect(props.setIsHourly).toHaveBeenCalledWith(43, 1, true);
  });

  test('handles changing the contractor yearly cost', () => {
    props.item.hourly.useHourly = false;
    const component = shallow(<ContractorForm {...props} />);

    component
      .findWhere(c => c.prop('name') === 'contractor-cost-ffy-1066')
      .simulate('change', { target: { value: 773 } });

    expect(props.setCostForYear).toHaveBeenCalledWith(43, 1, '1066', 773);
  });

  test('handles changing the contractor yearly cost', () => {
    props.item.hourly.useHourly = true;
    const parentComponent = shallow(<ContractorForm {...props} />);
    const component = shallow(
      parentComponent
        .findWhere(
          c => c.name() === 'ChoiceComponent' && c.prop('value') === 'yes'
        )
        .prop('checkedChildren')
    );

    component
      .findWhere(c => c.prop('name') === 'contractor-num-hours-ffy-1066')
      .simulate('change', { target: { value: 3752 } });

    expect(props.setNumberOfHoursForYear).toHaveBeenCalledWith(
      43,
      1,
      '1066',
      3752
    );
  });

  test('handles changing the contractor yearly cost', () => {
    props.item.hourly.useHourly = true;
    const parentComponent = shallow(<ContractorForm {...props} />);
    const component = shallow(
      parentComponent
        .findWhere(
          c => c.name() === 'ChoiceComponent' && c.prop('value') === 'yes'
        )
        .prop('checkedChildren')
    );

    component
      .findWhere(c => c.prop('name') === 'contractor-hourly-rate-ffy-1067')
      .simulate('change', { target: { value: 9364 } });

    expect(props.setHourlyRateForYear).toHaveBeenCalledWith(
      43,
      1,
      '1067',
      9364
    );
  });
});
