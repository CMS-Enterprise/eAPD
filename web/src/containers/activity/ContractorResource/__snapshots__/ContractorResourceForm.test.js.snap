// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the ContractorResourceForm component renders correctly 1`] = `
<Fragment>
  <TextField
    autoFocus={true}
    label="Contractor Name"
    name="contractor-name"
    type="text"
    value="Honey Dipper Dan"
  />
  <TextArea
    label="Description of Services"
    name="contractor-description"
    rows={5}
    value="They cleaned up the latrines after the Battle of Hastings"
  />
  <FormLabel
    component="label"
  >
    Contract Term
  </FormLabel>
  <div
    className="ds-c-choice__checkedChild ds-u-padding-y--0"
  >
    <DateField
      label="Start"
      onChange={[Function]}
      value="1066-10-14"
    />
    <DateField
      hint=""
      label="End"
      onChange={[Function]}
      value="1066-10-15"
    />
  </div>
  <DollarField
    label="Total Contract Cost"
    name="contractor-total-cost"
    onBlur={null}
    onChange={null}
    size="medium"
    value={12345}
  />
  <fieldset
    className="ds-c-fieldset"
  >
    <legend
      className="ds-c-label"
    >
      This is an hourly resource
    </legend>
    <ChoiceComponent
      checked={true}
      name="apd-activity-contractor-hourly-key 1-no"
      type="radio"
      value="no"
    >
      No
    </ChoiceComponent>
    <ChoiceComponent
      checked={false}
      checkedChildren={
        <div
          className="ds-c-choice__checkedChild"
        >
          <React.Fragment>
            <FormLabel
              component="label"
            >
              FFY 
              1066
            </FormLabel>
            <div
              className="ds-l-row ds-u-padding-left--2"
            >
              <NumberField
                fieldRef={null}
                label="Number of hours"
                labelClassName="ds-u-margin-top--1"
                name="contractor-num-hours-ffy-1066"
                size="medium"
                type="number"
                value={10}
              />
              <DollarField
                className="ds-u-margin-left--1"
                label="Hourly rate"
                labelClassName="ds-u-margin-top--1"
                name="contractor-hourly-rate-ffy-1066"
                onBlur={null}
                onChange={null}
                size="medium"
                value={100}
              />
            </div>
          </React.Fragment>
          <React.Fragment>
            <FormLabel
              component="label"
            >
              FFY 
              1067
            </FormLabel>
            <div
              className="ds-l-row ds-u-padding-left--2"
            >
              <NumberField
                fieldRef={null}
                label="Number of hours"
                labelClassName="ds-u-margin-top--1"
                name="contractor-num-hours-ffy-1067"
                size="medium"
                type="number"
                value={20}
              />
              <DollarField
                className="ds-u-margin-left--1"
                label="Hourly rate"
                labelClassName="ds-u-margin-top--1"
                name="contractor-hourly-rate-ffy-1067"
                onBlur={null}
                onChange={null}
                size="medium"
                value={200}
              />
            </div>
          </React.Fragment>
        </div>
      }
      name="apd-activity-contractor-hourly-key 1-yes"
      type="radio"
      value="yes"
    >
      Yes
    </ChoiceComponent>
  </fieldset>
  <DollarField
    disabled={false}
    key="1066"
    label="FFY 1066 Cost"
    name="contractor-cost-ffy-1066"
    onBlur={null}
    onChange={null}
    size="medium"
    value={300}
  />
  <DollarField
    disabled={false}
    key="1067"
    label="FFY 1067 Cost"
    name="contractor-cost-ffy-1067"
    onBlur={null}
    onChange={null}
    size="medium"
    value={400}
  />
</Fragment>
`;

exports[`the ContractorResourceForm component renders if use hourly data is selected 1`] = `
<Fragment>
  <TextField
    autoFocus={true}
    label="Contractor Name"
    name="contractor-name"
    type="text"
    value="Honey Dipper Dan"
  />
  <TextArea
    label="Description of Services"
    name="contractor-description"
    rows={5}
    value="They cleaned up the latrines after the Battle of Hastings"
  />
  <FormLabel
    component="label"
  >
    Contract Term
  </FormLabel>
  <div
    className="ds-c-choice__checkedChild ds-u-padding-y--0"
  >
    <DateField
      label="Start"
      onChange={[Function]}
      value="1066-10-14"
    />
    <DateField
      hint=""
      label="End"
      onChange={[Function]}
      value="1066-10-15"
    />
  </div>
  <DollarField
    label="Total Contract Cost"
    name="contractor-total-cost"
    onBlur={null}
    onChange={null}
    size="medium"
    value={12345}
  />
  <fieldset
    className="ds-c-fieldset"
  >
    <legend
      className="ds-c-label"
    >
      This is an hourly resource
    </legend>
    <ChoiceComponent
      checked={false}
      name="apd-activity-contractor-hourly-key 1-no"
      type="radio"
      value="no"
    >
      No
    </ChoiceComponent>
    <ChoiceComponent
      checked={true}
      checkedChildren={
        <div
          className="ds-c-choice__checkedChild"
        >
          <React.Fragment>
            <FormLabel
              component="label"
            >
              FFY 
              1066
            </FormLabel>
            <div
              className="ds-l-row ds-u-padding-left--2"
            >
              <NumberField
                fieldRef={null}
                label="Number of hours"
                labelClassName="ds-u-margin-top--1"
                name="contractor-num-hours-ffy-1066"
                size="medium"
                type="number"
                value={10}
              />
              <DollarField
                className="ds-u-margin-left--1"
                label="Hourly rate"
                labelClassName="ds-u-margin-top--1"
                name="contractor-hourly-rate-ffy-1066"
                onBlur={null}
                onChange={null}
                size="medium"
                value={100}
              />
            </div>
          </React.Fragment>
          <React.Fragment>
            <FormLabel
              component="label"
            >
              FFY 
              1067
            </FormLabel>
            <div
              className="ds-l-row ds-u-padding-left--2"
            >
              <NumberField
                fieldRef={null}
                label="Number of hours"
                labelClassName="ds-u-margin-top--1"
                name="contractor-num-hours-ffy-1067"
                size="medium"
                type="number"
                value={20}
              />
              <DollarField
                className="ds-u-margin-left--1"
                label="Hourly rate"
                labelClassName="ds-u-margin-top--1"
                name="contractor-hourly-rate-ffy-1067"
                onBlur={null}
                onChange={null}
                size="medium"
                value={200}
              />
            </div>
          </React.Fragment>
        </div>
      }
      name="apd-activity-contractor-hourly-key 1-yes"
      type="radio"
      value="yes"
    >
      Yes
    </ChoiceComponent>
  </fieldset>
  <p
    className="ds-u-margin-bottom--0"
  >
    <FormLabel
      component="label"
    >
      FFY 
      1066
       Cost
    </FormLabel>
    <Dollars
      long={false}
    >
      300
    </Dollars>
    <FormLabel
      component="label"
    >
      FFY 
      1067
       Cost
    </FormLabel>
    <Dollars
      long={false}
    >
      400
    </Dollars>
  </p>
</Fragment>
`;
